### ISSUE TITLE
"not enough elements in stack" assertion when playing with coroutines

### ISSUE THREAD

**Issue by jonas-schievink** [chucklefish/rlua#8]
Hit this while writing docs and an example for `LuaThread::resume`. Take this code:

```rust
extern crate rlua;
use rlua::*;
                                                             
fn main() {
let lua = Lua::new();
let thread: LuaThread = lua.eval(r#"
    coroutine.create(function(arg)
        assert(arg == 42)
        local yieldarg = coroutine.yield(123)
        assert(yieldarg == 43)
        return 987
    end)
"#).unwrap();
                                                             
assert_eq!(thread.resume::<_, u32>(42).unwrap(), Some(123));
assert_eq!(thread.resume::<_, u32>(43).unwrap(), Some(987));
assert_eq!(thread.resume::<_, u32>(()).unwrap(), None);
}
```

Executing it hits this assertion inside Lua:
```
rust_out: lua/ldo.c:663: lua_resume: Assertion `(((L->status == 0) ? nargs + 1 : nargs) < (L->top - L->ci->func)) && "not enough elements in the stack"' failed.
```

**Comment by kyren** [comment]
Yeah, crashes there don't surprise me, good catch!  I'll definitely try and fix that tonight.

**Comment by kyren** [comment]
Fixed by 7b00ff0, it wasn't even a deep bug I just skipped the part where I actually checked the status of the coroutine before attempting to resume it :/

@jonas-schievink, what do you think about the thread API here, I copied the API for threads from starbound's c++ lua api (which returns a Maybe<R>), but maybe this doesn't map to rust very well?

I'm thinking about changing the API to have a CoroutineDead error and just returning that, that makes more sense to me.  What do you think?

**Comment by jonas-schievink** [comment]
That's probably a good idea. APIs that nest option and result can get confusing fast.
